File: leetcode/python/670.maximum-swap.py
class Solution:
    def maximumSwap(self, num: int) -> int:
        # nums = list(map(int, list(str(num))))
        nums = list(str(num))
        changed = False
        ans = num
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                if nums[j] > nums[i]:
                    numsc = [i for i in nums]
                    numsc[i], numsc[j] = numsc[j], numsc[i]
                    x = int("".join(numsc))
                    if x > ans:
                        ans = x
        return ans

""" """

class Solution:
    def maximumSwap(self, num: int) -> int:
        s = list(str(num))
        position = {c: i for i, c in enumerate(s)}

        for i, c in enumerate(s):
            for digit in reversed(string.digits):
                if digit <= c:
                    break
                if digit in position and position[digit] > i:
                    s[i], s[position[digit]] = digit, s[i]
                    return int("".join(s))

        return num

File: leetcode/go/670.maximum-swap.go
func maximumSwap(num int) int {
    nums := []rune(strconv.Itoa(num))
    lastIdx := [10]int{-1, -1, -1, -1, -1, -1, -1, -1, -1, -1}
    for i, x := range nums {
        lastIdx[x - '0'] = i
    }
    for i:=0; i < len(nums); i++ {
        for j:=9; j > int(nums[i]-'0'); j-- {
            if lastIdx[j] > i {
                nums[lastIdx[j]], nums[i] = nums[i], nums[lastIdx[j]]
                // fmt.Println()
                ans, _ := strconv.Atoi(string(nums))
                return ans
            }
        }
    }
    return num
}

